HW #4

ChereP721 opened the pull request:
## Домашнее задание №4 «LRU-кэш»

### Критерии оценки
- [x] Пайплайн зелёный - 4 балла
- [ ] Добавлены новые юнит-тесты для списка - 1 балл
- [x] Добавлены новые юнит-тесты для кэша (включая тест на логику
выталкивания из кэша редко используемых элементов) - до 3 баллов
- [x] Понятность и чистота кода - до 2 баллов

#### Зачёт от 7 баллов

ChereP721 committed:
HW #4 version 1

ChereP721 committed:
HW #4 version 2
 - указатель заменен на значение

ChereP721 committed:
HW #4 version 3
 - в очереди лежат только ключи
 - значения в мапе

Antonboom commented on hw04_lru_cache/list.go:
```diff
@@ -1,15 +1,100 @@
+}
+
+func (l *list) Remove(i *listItem) {
+	// неплохо бы проверить, что элемент является членом этого списка, т.к. в задании сказано, что "Гарантируется, что методы Remove и MoveToFront вызываются от существующих в списке элементов." - пропускаем
```
именно :)

Antonboom commented on hw04_lru_cache/list.go:
```diff
@@ -1,15 +1,100 @@
 
 type List interface {
-	// Place your code here
+	Len() int                          // длина списка
```
а где новые тесты на список?

ChereP721 replied:
добавил

Antonboom commented on hw04_lru_cache/cache_test.go:
```diff
@@ -49,14 +59,121 @@ func TestCache(t *testing.T) {
+	})
+
+	/*
+		в задании сказано тест "на логику выталкивания редкоиспользуемых элементов"
```
вас устроит, если я заменю "редкоиспользуемых" на "давно используемых" ?

ChereP721 replied:
да

Antonboom replied:
https://github.com/OtusGolang/home_work/commit/657d901a6c776148a0ce9f75a093a015677bdea0

Antonboom commented on hw04_lru_cache/cache.go:
```diff
@@ -1,22 +1,86 @@
+		mu:       sync.Mutex{},
+		capacity: capacity,
+		queue:    NewList(),
+		items:    make(map[Key]*cacheItem),
```
преаллокация, capacity же имеем

ChereP721 replied:
косяк, исправил

Antonboom commented on hw04_lru_cache/cache.go:
```diff
@@ -1,22 +1,86 @@
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+
+	for key, cacheEl := range cache.items {
```
```
cache.queue = NewList()
cache.items = ...
```
?

ChereP721 replied:
переделал, а такой метод не дороже по памяти?

Antonboom replied:
скорее по нагрузке на GC, да

зависит от ситуации, просто имейте в виду, что  можно по-разному :)

ChereP721 replied:
ок

Antonboom commented on hw04_lru_cache/cache.go:
```diff
@@ -1,22 +1,86 @@
+	queueItem *listItem
+}
+
+func (cache *lruCache) Set(key Key, value interface{}) bool {
```
FYI: очень длинное название ресивера, непривычно для Go

обычно `c *lruCache`, `lc *lruCache`, etc.

ChereP721 replied:
принято, поменял

Antonboom commented on hw04_lru_cache/cache.go:
```diff
@@ -1,22 +1,86 @@
+		return true
+	}
+
+	if len(cache.items) == cache.capacity {
```
можно вынести в какой-нибудь приватный метод `purge`

ChereP721 replied:
вынес

Antonboom commented on hw04_lru_cache/cache.go:
```diff
@@ -1,22 +1,86 @@
+		delKey, ok := cache.queue.Back().Value.(Key)
+		if !ok {
+			// поскольку в интерфейсе нет возврата ошибки - паникуем
+			panic("Wrong cache item type!")
```
я бы предложил паниковать с более очевидным для разработчика сообщением, например,
```go
panic(fmt.Sprintf("wrong lruCache.Set realization: unexpected queue value type: %T", ...)
```

ChereP721 replied:
скорректировал

Antonboom reviewed (dismissed):
https://otus.ru/teacher-lk/homework/145399/9018/

Здравствуйте!
Отличная работа.

Принято
9 / 10

ChereP721 committed:
HW #4 correct by review
 - доп тесты для list и тест на очистку кеша
 - преаллокация для мапы
 - очистка через создание новых объектов
 - укоротил название ресиверов
 - выделен метод выталкивания pushOut
 - изменен текст паники

Antonboom reviewed (approved):
10 / 10
